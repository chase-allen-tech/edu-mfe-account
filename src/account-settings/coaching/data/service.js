import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { getConfig } from '@edx/frontend-platform';
import get from 'lodash.get';

/**
 * get all settings related to the coaching plugin. Settings used
 * by Microbachelors students.
 * @param {Number} userId users are identified in the api by LMS id
 */
export async function getCoachingPreferences(userId) {
  let data = {};
  try {
    ({ data } = await getAuthenticatedHttpClient()
      .get(`${getConfig().LMS_BASE_URL}/api/coaching/v1/users/${userId}/`));
  } catch (error) {
    // If a user isn't active the API call will fail with a lack of credentials.
    data = {
      coaching_consent: false,
      user: userId,
      eligible_for_coaching: false,
      consent_form_seen: false,
    };
  }

  return data;
}

/**
 * patch all of the settings related to coaching.
 * @param {Number} userId users are identified in the api by LMS id
 * @param {Object} commitValues { coaching }
 */
export async function patchCoachingPreferences(userId, commitValues) {
  const requestUrl = `${getConfig().LMS_BASE_URL}/api/coaching/v1/users/${userId}/`;
  const { coaching } = commitValues;
  coaching.user = userId;

  await getAuthenticatedHttpClient()
    .patch(requestUrl, coaching)
    .catch((error) => {
      const apiError = Object.create(error);
      apiError.fieldErrors = JSON.parse(error.customAttributes.httpErrorResponseData);
      if (get(apiError, 'fieldErrors.phone_number')) {
        // eslint-disable-next-line prefer-destructuring
        apiError.fieldErrors.coaching = apiError.fieldErrors.phone_number[0];
        delete apiError.fieldErrors.phone_number;
      }
      throw apiError;
    });
  return commitValues;
}
