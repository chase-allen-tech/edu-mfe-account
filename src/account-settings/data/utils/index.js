export {
  AsyncActionType,
  getModuleState,
} from './reduxUtils';
export { default as handleFailure } from './sagaUtils';
export { unpackFieldErrors, handleRequestError } from './serviceUtils';
