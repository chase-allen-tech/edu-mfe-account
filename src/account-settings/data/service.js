import { getConfig } from '@edx/frontend-platform';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import pick from 'lodash.pick';
import pickBy from 'lodash.pickby';
import omit from 'lodash.omit';
import isEmpty from 'lodash.isempty';

import { handleRequestError, unpackFieldErrors } from './utils';
import { getThirdPartyAuthProviders } from '../third-party-auth';
import { getCoachingPreferences, patchCoachingPreferences } from '../coaching/data/service';
import { getDemographics, getDemographicsOptions, patchDemographics } from '../demographics/data/service';
import { DEMOGRAPHICS_FIELDS } from '../demographics/data/utils';

const SOCIAL_PLATFORMS = [
  { id: 'twitter', key: 'social_link_twitter' },
  { id: 'facebook', key: 'social_link_facebook' },
  { id: 'linkedin', key: 'social_link_linkedin' },
];

function unpackAccountResponseData(data) {
  const unpackedData = data;

  // This is handled by preferences
  delete unpackedData.time_zone;

  SOCIAL_PLATFORMS.forEach(({ id, key }) => {
    const platformData = data.social_links.find(({ platform }) => platform === id);
    unpackedData[key] = typeof platformData === 'object' ? platformData.social_link : '';
  });

  if (Array.isArray(data.language_proficiencies)) {
    if (data.language_proficiencies.length) {
      unpackedData.language_proficiencies = data.language_proficiencies[0].code;
    } else {
      unpackedData.language_proficiencies = '';
    }
  }

  return unpackedData;
}

function packAccountCommitData(commitData) {
  const packedData = commitData;

  SOCIAL_PLATFORMS.forEach(({ id, key }) => {
    // Skip missing values. Empty strings are valid values and should be preserved.
    if (commitData[key] === undefined) {
      return;
    }

    packedData.social_links = [{ platform: id, social_link: commitData[key] }];
    delete packedData[key];
  });

  if (commitData.language_proficiencies !== undefined) {
    if (commitData.language_proficiencies) {
      packedData.language_proficiencies = [{ code: commitData.language_proficiencies }];
    } else {
      // An empty string should be sent as an array.
      packedData.language_proficiencies = [];
    }
  }

  if (commitData.year_of_birth !== undefined) {
    if (commitData.year_of_birth) {
      packedData.year_of_birth = commitData.year_of_birth;
    } else {
      // An empty string should be sent as null.
      packedData.year_of_birth = null;
    }
  }
  return packedData;
}

export async function getAccount(username) {
  const { data } = await getAuthenticatedHttpClient()
    .get(`${getConfig().LMS_BASE_URL}/api/user/v1/accounts/${username}`);
  return unpackAccountResponseData(data);
}

export async function patchAccount(username, commitValues) {
  const requestConfig = {
    headers: { 'Content-Type': 'application/merge-patch+json' },
  };

  const { data } = await getAuthenticatedHttpClient()
    .patch(
      `${getConfig().LMS_BASE_URL}/api/user/v1/accounts/${username}`,
      packAccountCommitData(commitValues),
      requestConfig,
    )
    .catch((error) => {
      const unpackFunction = (fieldErrors) => {
        const unpackedFieldErrors = fieldErrors;
        if (fieldErrors.social_links) {
          SOCIAL_PLATFORMS.forEach(({ key }) => {
            unpackedFieldErrors[key] = fieldErrors.social_links;
          });
        }
        return unpackFieldErrors(unpackedFieldErrors);
      };
      handleRequestError(error, unpackFunction);
    });

  return unpackAccountResponseData(data);
}

export async function getPreferences(username) {
  const { data } = await getAuthenticatedHttpClient()
    .get(`${getConfig().LMS_BASE_URL}/api/user/v1/preferences/${username}`);
  return data;
}

export async function patchPreferences(username, commitValues) {
  const requestConfig = { headers: { 'Content-Type': 'application/merge-patch+json' } };
  const requestUrl = `${getConfig().LMS_BASE_URL}/api/user/v1/preferences/${username}`;

  // Ignore the success response, the API does not currently return any data.
  await getAuthenticatedHttpClient()
    .patch(requestUrl, commitValues, requestConfig).catch(handleRequestError);

  return commitValues;
}

export async function getTimeZones(forCountry) {
  const { data } = await getAuthenticatedHttpClient()
    .get(`${getConfig().LMS_BASE_URL}/user_api/v1/preferences/time_zones/`, {
      params: { country_code: forCountry },
    })
    .catch(handleRequestError);

  return data;
}

/**
 * Determine if the user's profile data is managed by a third-party identity provider.
 */
export async function getProfileDataManager(username, userRoles) {
  const userRoleNames = userRoles.map(role => role.split(':')[0]);

  if (userRoleNames.includes('enterprise_learner')) {
    const url = `${getConfig().LMS_BASE_URL}/enterprise/api/v1/enterprise-learner/?username=${username}`;
    const { data } = await getAuthenticatedHttpClient().get(url).catch(handleRequestError);

    if (data.results.length > 0) {
      const enterprise = data.results[0] && data.results[0].enterprise_customer;
      // To ensure that enterprise returned is current enterprise & it manages profile settings
      if (enterprise && enterprise.sync_learner_profile_data) {
        return enterprise.name;
      }
    }
  }

  return null;
}

/**
 * A function to determine if the Demographics questions should be displayed to the user. For the
 * MVP release of Demographics we are limiting the Demographics question visibility only to
 * MicroBachelors learners.
 */
export async function shouldDisplayDemographicsQuestions() {
  const requestUrl = `${getConfig().LMS_BASE_URL}/api/demographics/v1/demographics/status/`;
  let data = {};

  try {
    ({ data } = await getAuthenticatedHttpClient().get(requestUrl));
    if (data.display) {
      return data.display;
    }
  } catch (error) {
    // if there was an error then we just hide the section
    return false;
  }

  return false;
}

/**
 * A single function to GET everything considered a setting.
 * Currently encapsulates Account, Preferences, Coaching, ThirdPartyAuth, and Demographics
 */
export async function getSettings(username, userRoles, userId) {
  const results = await Promise.all([
    getAccount(username),
    getPreferences(username),
    getThirdPartyAuthProviders(),
    getProfileDataManager(username, userRoles),
    getTimeZones(),
    getConfig().COACHING_ENABLED && getCoachingPreferences(userId),
    getConfig().ENABLE_DEMOGRAPHICS_COLLECTION && shouldDisplayDemographicsQuestions(),
    getConfig().ENABLE_DEMOGRAPHICS_COLLECTION && getDemographics(userId),
    getConfig().ENABLE_DEMOGRAPHICS_COLLECTION && getDemographicsOptions(),
  ]);

  return {
    ...results[0],
    ...results[1],
    thirdPartyAuthProviders: results[2],
    profileDataManager: results[3],
    timeZones: results[4],
    coaching: results[5],
    shouldDisplayDemographicsSection: results[6],
    ...results[7], // demographics
    demographicsOptions: results[8],
  };
}

/**
 * A single function to PATCH everything considered a setting.
 * Currently encapsulates Account, Preferences, coaching and ThirdPartyAuth
 */
export async function patchSettings(username, commitValues, userId) {
  // Note: time_zone exists in the return value from user/v1/accounts
  // but it is always null and won't update. It also exists in
  // user/v1/preferences where it does update. This is the one we use.
  const preferenceKeys = ['time_zone'];
  const coachingKeys = ['coaching'];
  const demographicsKeys = DEMOGRAPHICS_FIELDS;
  const isDemographicsKey = (value, key) => key.includes('demographics');
  const accountCommitValues = omit(commitValues, preferenceKeys, coachingKeys, demographicsKeys);
  const preferenceCommitValues = pick(commitValues, preferenceKeys);
  const coachingCommitValues = pick(commitValues, coachingKeys);
  const demographicsCommitValues = pickBy(commitValues, isDemographicsKey);
  const patchRequests = [];

  if (!isEmpty(accountCommitValues)) {
    patchRequests.push(patchAccount(username, accountCommitValues));
  }
  if (!isEmpty(preferenceCommitValues)) {
    patchRequests.push(patchPreferences(username, preferenceCommitValues));
  }
  if (!isEmpty(coachingCommitValues)) {
    patchRequests.push(patchCoachingPreferences(userId, coachingCommitValues));
  }
  if (!isEmpty(demographicsCommitValues)) {
    patchRequests.push(patchDemographics(userId, demographicsCommitValues));
  }

  const results = await Promise.all(patchRequests);
  // Assigns in order of requests. Preference keys
  // will override account keys. Notably time_zone.
  const combinedResults = Object.assign({}, ...results);
  return combinedResults;
}
