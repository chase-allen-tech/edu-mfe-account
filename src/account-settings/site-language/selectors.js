import { createSelector } from 'reselect';
import { getModuleState } from '../data/utils';

export const storePath = ['accountSettings', 'siteLanguage'];

const siteLanguageSelector = state => getModuleState(state, storePath);

export const siteLanguageListSelector = createSelector(
  siteLanguageSelector,
  siteLanguage => siteLanguage.siteLanguageList,
);

export const siteLanguageOptionsSelector = createSelector(
  siteLanguageSelector,
  siteLanguage => siteLanguage.siteLanguageList.map(({ code, name }) => ({
    value: code,
    label: name,
  })),
);
