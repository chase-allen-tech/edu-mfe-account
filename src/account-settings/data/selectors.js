import { createSelector, createStructuredSelector } from 'reselect';
import { siteLanguageListSelector, siteLanguageOptionsSelector } from '../site-language';

export const storeName = 'accountSettings';

export const accountSettingsSelector = state => ({ ...state[storeName] });

const editableFieldNameSelector = (state, props) => props.name;

const valuesSelector = createSelector(
  accountSettingsSelector,
  accountSettings => accountSettings.values,
);

const draftsSelector = createSelector(
  accountSettingsSelector,
  accountSettings => accountSettings.drafts,
);

const previousSiteLanguageSelector = createSelector(
  accountSettingsSelector,
  accountSettings => accountSettings.previousSiteLanguage,
);

const editableFieldErrorSelector = createSelector(
  editableFieldNameSelector,
  accountSettingsSelector,
  (name, accountSettings) => accountSettings.errors[name],
);

const editableFieldConfirmationValuesSelector = createSelector(
  editableFieldNameSelector,
  accountSettingsSelector,
  (name, accountSettings) => accountSettings.confirmationValues[name],
);

const isEditingSelector = createSelector(
  editableFieldNameSelector,
  accountSettingsSelector,
  (name, accountSettings) => accountSettings.openFormId === name,
);

const confirmationValuesSelector = createSelector(
  accountSettingsSelector,
  accountSettings => accountSettings.confirmationValues,
);

const errorSelector = createSelector(
  accountSettingsSelector,
  accountSettings => accountSettings.errors,
);

const saveStateSelector = createSelector(
  accountSettingsSelector,
  accountSettings => accountSettings.saveState,
);

export const editableFieldSelector = createStructuredSelector({
  error: editableFieldErrorSelector,
  confirmationValue: editableFieldConfirmationValuesSelector,
  saveState: saveStateSelector,
  isEditing: isEditingSelector,
});

export const profileDataManagerSelector = createSelector(
  accountSettingsSelector,
  accountSettings => accountSettings.profileDataManager,
);

export const staticFieldsSelector = createSelector(
  accountSettingsSelector,
  accountSettings => (accountSettings.profileDataManager ? ['name', 'email', 'country'] : []),
);

/**
 * If there's no draft present at all (undefined), use the original committed value.
 */
function chooseFormValue(draft, committed) {
  return draft !== undefined ? draft : committed;
}

const formValuesSelector = createSelector(
  valuesSelector,
  draftsSelector,
  (values, drafts) => {
    const formValues = {};
    Object.entries(values).forEach(([name, value]) => {
      formValues[name] = chooseFormValue(drafts[name], value) || '';
    });
    return formValues;
  },
);

const transformTimeZonesToOptions = timeZoneArr => timeZoneArr
  .map(({ time_zone, description }) => ({ // eslint-disable-line camelcase
    value: time_zone, label: description,
  }));

const timeZonesSelector = createSelector(
  accountSettingsSelector,
  accountSettings => transformTimeZonesToOptions(accountSettings.timeZones),
);

const countryTimeZonesSelector = createSelector(
  accountSettingsSelector,
  accountSettings => transformTimeZonesToOptions(accountSettings.countryTimeZones),
);

const activeAccountSelector = createSelector(
  accountSettingsSelector,
  accountSettings => accountSettings.values.is_active,
);

export const siteLanguageSelector = createSelector(
  previousSiteLanguageSelector,
  draftsSelector,
  (previousValue, drafts) => ({
    previousValue,
    draft: drafts.siteLanguage,
  }),
);

export const betaLanguageBannerSelector = createStructuredSelector({
  siteLanguageList: siteLanguageListSelector,
  siteLanguage: siteLanguageSelector,
});

export const accountSettingsPageSelector = createSelector(
  accountSettingsSelector,
  siteLanguageOptionsSelector,
  siteLanguageSelector,
  formValuesSelector,
  profileDataManagerSelector,
  staticFieldsSelector,
  timeZonesSelector,
  countryTimeZonesSelector,
  activeAccountSelector,
  saveStateSelector,
  (
    accountSettings,
    siteLanguageOptions,
    siteLanguage,
    formValues,
    profileDataManager,
    staticFields,
    timeZoneOptions,
    countryTimeZoneOptions,
    activeAccount,
    saveState,
  ) => ({
    siteLanguageOptions,
    siteLanguage,
    loading: accountSettings.loading,
    loaded: accountSettings.loaded,
    loadingError: accountSettings.loadingError,
    timeZoneOptions,
    countryTimeZoneOptions,
    isActive: activeAccount,
    formValues,
    profileDataManager,
    staticFields,
    tpaProviders: accountSettings.thirdPartyAuth.providers,
    saveState,
  }),
);

export const coachingConsentPageSelector = createSelector(
  accountSettingsSelector,
  formValuesSelector,
  activeAccountSelector,
  profileDataManagerSelector,
  saveStateSelector,
  confirmationValuesSelector,
  errorSelector,
  (
    accountSettings,
    formValues,
    activeAccount,
    profileDataManager,
    saveState,
    confirmationValues,
    errors,
  ) => ({
    loading: accountSettings.loading,
    loaded: accountSettings.loaded,
    loadingError: accountSettings.loadingError,
    isActive: activeAccount,
    profileDataManager,
    formValues,
    saveState,
    confirmationValues,
    formErrors: errors,
  }),
);

export const demographicsSectionSelector = createSelector(
  formValuesSelector,
  draftsSelector,
  errorSelector,
  (
    formValues,
    drafts,
    errors,
  ) => ({
    formValues,
    drafts,
    formErrors: errors,
  }),
);
