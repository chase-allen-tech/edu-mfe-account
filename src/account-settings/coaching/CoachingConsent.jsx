import React from 'react';

import { getConfig, getQueryParameters } from '@edx/frontend-platform';
import { injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { Hyperlink } from '@edx/paragon';
import { faCheck } from '@fortawesome/free-solid-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import get from 'lodash.get';
import { getAuthenticatedHttpClient, getAuthenticatedUser } from '@edx/frontend-platform/auth';

import PageLoading from '../PageLoading';
import CoachingConsentForm from './CoachingConsentForm';
import messages from './CoachingConsent.messages';
import LogoSVG from '../../logo.svg';
import { fetchSettings } from '../data/actions';
import { coachingConsentPageSelector } from '../data/selectors';

const Logo = ({ src, alt, ...attributes }) => (
  <>
    <img src={src} alt={alt} {...attributes} />
  </>
);

const SuccessMessage = props => (
  <div className="col-12 col-lg-6 shadow-lg mx-auto mt-4 p-5">
    <FontAwesomeIcon className="text-success" icon={faCheck} size="5x" />
    <div className="h3">{props.header}</div>
    <div>{props.message}</div>
    <Hyperlink destination={props.continueUrl} className="d-block p-2 my-3 text-center text-white bg-primary rounded">
      {props.continue}
    </Hyperlink>
  </div>
);

const AutoRedirect = (props) => {
  window.location.href = props.redirectUrl;
  return <></>;
};

const VIEWS = {
  NOT_LOADED: 'NOT_LOADED',
  LOADED: 'LOADED',
  SUCCESS: 'SUCCESS',
  SUCCESS_PENDING: 'SUCCESS_PENDING',
  DECLINED: 'DECLINED',
  DECLINE_PENDING: 'DECLINE_PENDING',
};

class CoachingConsent extends React.Component {
  constructor(props, context) {
    super(props, context);

    // Used to redirect back to the courseware.
    const nextUrl = this.sanitizeForwardingUrl(getQueryParameters().next);
    this.state = {
      redirectUrl: nextUrl || `${getConfig().LMS_BASE_URL}/dashboard/`,
      formErrors: {},
      formSubmitted: false,
      declineSubmitted: false,
      submissionSuccess: false,
    };

    this.handleSubmit = this.handleSubmit.bind(this);
    this.declineCoaching = this.declineCoaching.bind(this);
    this.patchUsingCoachingConsentForm = this.patchUsingCoachingConsentForm.bind(this);
  }

  componentDidMount() {
    this.props.fetchSettings();
  }

  sanitizeForwardingUrl(url) {
    // Redirect to root of MFE if invalid next param is sent
    return url && url.startsWith(getConfig().LMS_BASE_URL) ? url : `${getConfig().LMS_BASE_URL}/dashboard/`;
  }

  async patchUsingCoachingConsentForm(body) {
    const { userId } = getAuthenticatedUser();
    const requestUrl = `${getConfig().LMS_BASE_URL}/api/coaching/v1/coaching_consent/${userId}/`;
    let formErrors = {};
    const data = await getAuthenticatedHttpClient()
      .patch(requestUrl, body)
      .catch((error) => {
        if (get(error, 'customAttributes.httpErrorResponseData')) {
          formErrors = JSON.parse(error.customAttributes.httpErrorResponseData);
        } else {
          formErrors = { full_name: 'Something went wrong. Please try again.' };
        }
        this.setState({
          submissionSuccess: false,
          formErrors,
          formSubmitted: false,
        });
      });
    if (get(data, 'status') === 200) {
      this.setState({ submissionSuccess: true });
    }
  }

  handleSubmit(e) {
    e.preventDefault();
    const fullName = e.target.fullName.value;
    const phoneNumber = e.target.phoneNumber.value;
    const body = {
      coaching_consent: true,
      consent_form_seen: true,
      phone_number: phoneNumber,
      full_name: fullName,
    };
    this.setState({
      formErrors: {},
      formSubmitted: true,
      declineSubmitted: false,
    }, () => this.patchUsingCoachingConsentForm(body));
  }

  declineCoaching(e) {
    e.preventDefault();
    const body = {
      coaching_consent: false,
      consent_form_seen: true,
    };
    this.setState({
      formErrors: {},
      formSubmitted: false,
      declineSubmitted: true,
    }, () => this.patchUsingCoachingConsentForm(body));
  }

  renderView(currentView) {
    switch (currentView) {
      case VIEWS.NOT_LOADED:
        return <PageLoading srMessage="" />;
      case VIEWS.LOADED:
        return (
          <CoachingConsentForm
            onSubmit={this.handleSubmit}
            declineCoaching={this.declineCoaching}
            formErrors={this.state.formErrors}
            formValues={this.props.formValues}
            redirectUrl={this.state.redirectUrl}
            profileDataManager={this.props.profileDataManager}
          />
        );
      case VIEWS.SUCCESS_PENDING:
        return <PageLoading srMessage="Submitting..." />;
      case VIEWS.SUCCESS:
        return (
          <SuccessMessage
            continueUrl={this.state.redirectUrl}
            header={this.props.intl.formatMessage(messages['account.settings.coaching.consent.success.header'])}
            message={this.props.intl.formatMessage(messages['account.settings.coaching.consent.success.message'])}
            continue={this.props.intl.formatMessage(messages['account.settings.coaching.consent.success.continue'])}
          />
        );
      case VIEWS.DECLINE_PENDING:
        return <PageLoading srMessage="Redirecting..." />;
      case VIEWS.DECLINED:
        return <AutoRedirect redirectUrl={this.state.redirectUrl} />;
      default:
        return <></>;
    }
  }

  render() {
    const { loaded } = this.props;
    const formHasErrors = Object.keys(this.state.formErrors).length > 0;
    let currentView = null;
    // This amount of logic was making the template very hard to read, so I broke it out into views.
    if (!loaded) {
      currentView = VIEWS.NOT_LOADED;
    } else if (this.state.formSubmitted && !formHasErrors) {
      if (this.state.submissionSuccess) {
        currentView = VIEWS.SUCCESS;
      } else {
        currentView = VIEWS.SUCCESS_PENDING;
      }
    } else if (this.state.declineSubmitted && !formHasErrors) {
      if (this.state.submissionSuccess) {
        currentView = VIEWS.DECLINED;
      } else {
        currentView = VIEWS.DECLINE_PENDING;
      }
    } else {
      currentView = VIEWS.LOADED;
    }

    return (
      <main>
        <div className="w-100 d-flex justify-content-center align-items-center shadow coaching-header">
          <Logo
            className="logo"
            src={LogoSVG}
            alt="Logo"
          />
        </div>
        {this.renderView(currentView)}
      </main>
    );
  }
}

Logo.defaultProps = {
  src: '',
  alt: '',
};

Logo.propTypes = {
  src: PropTypes.string,
  alt: PropTypes.string,
};

SuccessMessage.defaultProps = {
  header: '',
  message: '',
  continueUrl: '',
  continue: '',
};

SuccessMessage.propTypes = {
  header: PropTypes.string,
  message: PropTypes.string,
  continueUrl: PropTypes.string,
  continue: PropTypes.string,
};

AutoRedirect.defaultProps = {
  redirectUrl: '',
};

AutoRedirect.propTypes = {
  redirectUrl: PropTypes.string,
};

CoachingConsent.defaultProps = {
  loaded: false,
  profileDataManager: null,
};

CoachingConsent.propTypes = {
  intl: intlShape.isRequired,
  loaded: PropTypes.bool,
  formValues: PropTypes.shape({
    name: PropTypes.string,
    phone_number: PropTypes.string,
    coaching: PropTypes.shape({
      coaching_consent: PropTypes.bool.isRequired,
      user: PropTypes.number.isRequired,
      eligible_for_coaching: PropTypes.bool.isRequired,
      consent_form_seen: PropTypes.bool.isRequired,
    }),
  }).isRequired,
  formErrors: PropTypes.shape({
    coaching: PropTypes.object,
  }).isRequired,
  confirmationValues: PropTypes.shape({
    coaching: PropTypes.object,
    name: PropTypes.object,
    phone_number: PropTypes.object,
  }).isRequired,
  fetchSettings: PropTypes.func.isRequired,
  profileDataManager: PropTypes.string,
};

export default connect(coachingConsentPageSelector, {
  fetchSettings,
})(injectIntl(CoachingConsent));
