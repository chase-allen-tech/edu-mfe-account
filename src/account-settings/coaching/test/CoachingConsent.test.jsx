import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { act } from 'react-dom/test-utils';
import configureStore from 'redux-mock-store';
import { IntlProvider, injectIntl } from '@edx/frontend-platform/i18n';
import * as auth from '@edx/frontend-platform/auth';

import CoachingConsent from '../CoachingConsent';
import * as selectors from '../../data/selectors';

jest.mock('@edx/frontend-platform/auth');

const IntlCoachingConsent = injectIntl(CoachingConsent);

jest.mock('../../data/selectors', () => jest.fn().mockImplementation(() => ({ coachingConsentPageSelector: () => ({}) })));

const mockStore = configureStore();

describe('CoachingConsent', () => {
  let props = {};
  let store = {};
  selectors.mockClear();

  const reduxWrapper = children => (
    <IntlProvider locale="en">
      <Provider store={store}>{children}</Provider>
    </IntlProvider>
  );

  beforeEach(() => {
    store = mockStore();
    props = {
      fetchSettings: jest.fn(),
      loaded: true,
      saveState: undefined,
      formValues: {
        name: 'edx edx',
        phone_number: '1234567890',
        coaching: {
          coaching_consent: true,
          consent_form_seen: false,
          eligible_for_coaching: true,
          user: 1,
        },
      },
      formErrors: {},
      confirmationValues: {},
      profileDataManager: '',
      intl: {},
    };
    auth.getAuthenticatedHttpClient = jest.fn(() => ({
      patch: async () => ({
        data: { status: 200 },
        catch: () => {},
      }),
    }));
    auth.getAuthenticatedUser = jest.fn(() => ({ userId: 3 }));
  });

  it('should render', () => {
    const wrapper = renderer.create(reduxWrapper(<IntlCoachingConsent {...props} />)).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('disables name field on enterprise user', () => {
    props = {
      ...props,
      profileDataManager: 'test person',
    };
    const wrapper = renderer.create(reduxWrapper(<IntlCoachingConsent {...props} />)).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('display completed box when successfully submitted', async () => {
    const fakeEvent = {
      preventDefault: () => {},
      target: {
        fullName: { value: 'edx edx' },
        phoneNumber: { value: '9783028731' },
      },
    };
    const wrapper = renderer.create(
      reduxWrapper(<IntlCoachingConsent {...props} />),
      {
        // bypass the forward-ref. we don't care about focus for this one test
        createNodeMock: (element) => {
          if (element.type === 'button') {
            // mock a focus function
            return {
              focus: async () => wrapper.root.findByType('form').props.onSubmit(fakeEvent),
            };
          }
          return null;
        },
      },
    );
    const form = wrapper.root.findByType('form');
    await act(async () => { await form.props.onSubmit(fakeEvent); });
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
