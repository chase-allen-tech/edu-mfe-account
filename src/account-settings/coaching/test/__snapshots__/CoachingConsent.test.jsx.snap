// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CoachingConsent disables name field on enterprise user 1`] = `
<main>
  <div
    className="w-100 d-flex justify-content-center align-items-center shadow coaching-header"
  >
    <img
      alt="Logo"
      className="logo"
      src="icon/mock/path"
    />
  </div>
  <div
    className="col-12 col-md-6 col-xl-5 mx-auto mt-4 p-5 shadow-lg"
  >
    <h2
      className="h2"
    >
      Let’s get started.
    </h2>
    <p>
      MicroBachelors programs include coaching that focuses on your career, education, and how you'll achieve results through one-on-one communication with an experienced professional. If you’re interested, provide the information below and click “Submit,” and our coaching partner will connect with you via email and/or text message to help you move forward. Terms and conditions apply.*
    </p>
    <div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="py-3"
        >
          <div
            className="alert d-flex align-items-start alert alert-primary"
          >
            <div />
            <div>
              <span>
                Your name is managed by 
                <b>
                  test person
                </b>
                . Contact your administrator for help.
              </span>
            </div>
          </div>
          <div
            className="alert-warning mb-2"
          >
            
          </div>
          <label
            className="h6"
            htmlFor="fullName"
          >
            Please confirm your name
          </label>
          <input
            className="form-control"
            defaultValue="edx edx"
            disabled={true}
            id="fullName"
            name="full-name"
            type="text"
          />
        </div>
        <div
          className="py-3"
        >
          <div
            className="alert-warning mb-2"
          >
            
          </div>
          <label
            className="h6"
            htmlFor="phoneNumber"
          >
            Enter your mobile number
          </label>
          <input
            className="form-control"
            defaultValue="1234567890"
            id="phoneNumber"
            name="phone_number"
            type="text"
          />
        </div>
        <div
          className=" py-3"
        >
          <p
            className="small font-italic"
          >
            * Coaching services are included at no additional cost to learners with US phone numbers. Coaching includes recurring text messages. Message and data rates may apply. Text STOP to opt-out.
          </p>
        </div>
        <div
          className="alert-warning mb-2"
        >
          
        </div>
        <div
          className="d-flex flex-column align-items-center"
        >
          <button
            className="w-100 btn btn-outline-primary"
            disabled={false}
            type="submit"
          >
            Sign up for coaching
          </button>
        </div>
        <div
          className="mt-3"
        >
          <a
            className="mt-3 text-dark btn-link small"
            href="http://localhost:18000/dashboard/"
            onClick={[Function]}
            target="_self"
          >
            I prefer not to be contacted with free coaching services
          </a>
        </div>
      </form>
    </div>
  </div>
</main>
`;

exports[`CoachingConsent display completed box when successfully submitted 1`] = `
<main>
  <div
    className="w-100 d-flex justify-content-center align-items-center shadow coaching-header"
  >
    <img
      alt="Logo"
      className="logo"
      src="icon/mock/path"
    />
  </div>
  <div>
    <div
      className="d-flex justify-content-center align-items-center flex-column"
      style={
        Object {
          "height": "50vh",
        }
      }
    >
      <div
        className="spinner-border text-primary"
        role="status"
      >
        <span
          className="sr-only"
        >
          Submitting...
        </span>
      </div>
    </div>
  </div>
</main>
`;

exports[`CoachingConsent should render 1`] = `
<main>
  <div
    className="w-100 d-flex justify-content-center align-items-center shadow coaching-header"
  >
    <img
      alt="Logo"
      className="logo"
      src="icon/mock/path"
    />
  </div>
  <div
    className="col-12 col-md-6 col-xl-5 mx-auto mt-4 p-5 shadow-lg"
  >
    <h2
      className="h2"
    >
      Let’s get started.
    </h2>
    <p>
      MicroBachelors programs include coaching that focuses on your career, education, and how you'll achieve results through one-on-one communication with an experienced professional. If you’re interested, provide the information below and click “Submit,” and our coaching partner will connect with you via email and/or text message to help you move forward. Terms and conditions apply.*
    </p>
    <div>
      <form
        onSubmit={[Function]}
      >
        <div
          className="py-3"
        >
          <div
            className="alert-warning mb-2"
          >
            
          </div>
          <label
            className="h6"
            htmlFor="fullName"
          >
            Please confirm your name
          </label>
          <input
            className="form-control"
            defaultValue="edx edx"
            disabled={false}
            id="fullName"
            name="full-name"
            type="text"
          />
        </div>
        <div
          className="py-3"
        >
          <div
            className="alert-warning mb-2"
          >
            
          </div>
          <label
            className="h6"
            htmlFor="phoneNumber"
          >
            Enter your mobile number
          </label>
          <input
            className="form-control"
            defaultValue="1234567890"
            id="phoneNumber"
            name="phone_number"
            type="text"
          />
        </div>
        <div
          className=" py-3"
        >
          <p
            className="small font-italic"
          >
            * Coaching services are included at no additional cost to learners with US phone numbers. Coaching includes recurring text messages. Message and data rates may apply. Text STOP to opt-out.
          </p>
        </div>
        <div
          className="alert-warning mb-2"
        >
          
        </div>
        <div
          className="d-flex flex-column align-items-center"
        >
          <button
            className="w-100 btn btn-outline-primary"
            disabled={false}
            type="submit"
          >
            Sign up for coaching
          </button>
        </div>
        <div
          className="mt-3"
        >
          <a
            className="mt-3 text-dark btn-link small"
            href="http://localhost:18000/dashboard/"
            onClick={[Function]}
            target="_self"
          >
            I prefer not to be contacted with free coaching services
          </a>
        </div>
      </form>
    </div>
  </div>
</main>
`;
